### PR TITLE
Track metadata abount peers on connection

### PR DESCRIPTION
**Pull request by pipermerriam**
fixes https://github.com/ethereum/trinity/issues/24

### What was wrong?

Every time the trinity client runs, it starts with zero knowledge about the network.  This means that it must bootstrap with the kademlia network and then find appropriate peers to connect to.  This process is time consuming and results in the node taking a long time to find appropriate peers and establish connections to start syncing.

### How was it fixed?

Now, each time we successfully connect to a peer, we keep a record of information about them.

- remote URI
- whether they were originally inbound or outbound
- when we first saw them
- when we last updated their record
- when we last established a connection with them that was at least X seconds long
- their genesis hash
- their network id
- the protocol name they were speaking
- the protocol version they were speaking

Now, when we start, the peer pool will query this database for nodes using the following rules.

- exclude nodes that have an active blacklist record.
- exclude nodes that were originally inbound
- exclude nodes who's network parameters do not match ours (genesis hash, network, etc)
- order nodes by the last time a good connection was established (most recent first)
- exclude nodes that we are already connected to

With this change, a node that goes offline and comes quickly back online will quickly re-establish many of their previous connections as they will be first in the queue.

[//]: # (For important changes, please add a new entry to the release notes file)
[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by pipermerriam** [comment]
Some *sub-tasks* I'm discovering that need to be pulled out.

- It's difficult to cause trinity shutdown from within an isolated scope plugin when the shutdown condition can only be detected within `do_start`.
- It's difficult to *properly* run more than one service in an isolated scope plugin since the utils in `trinity._utils.shutdown` only support singular services.

File: p2p/peer_pool.py [hunk 1]
```diff
@@ -395,11 +396,22 @@ def _add_peer(self,
...
+            end_at = time.perf_counter()
+            self.logger.debug(
+                "Failed attempt to connect to %s took %s seconds",
+                node,
+                end_at - start_at,
+            )
             return
+        else:
+            end_at = time.perf_counter()
+            self.logger.debug("Successful connect to %s took %s seconds", node, end_at - start_at)
```
Comment (pipermerriam) [inline p2p/peer_pool.py]: TODO, remove this new timing code.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+        self.update_peer_meta(
+            remote=peer.remote,
+            genesis_hash=peer.genesis_hash,
+            protocol=peer.sub_proto.name,
+            protocol_version=peer.sub_proto.version,
+            network_id=peer.network_id,
+        )
+
+    def deregister_peer(self, peer: BasePeer) -> None:
+        # TODO: this is where we can track actual performance data about peers
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: Remove `TODO`

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        if self._meta_exists(uri):
+            self.logger.debug("Updated metadata: %s", remote)
+            eth1_meta = self._get_meta(uri)
+            eth1_meta.genesis_hash = genesis_hash.hex()
+            eth1_meta.protocol = protocol
+            eth1_meta.protocol_version = protocol_version
+            eth1_meta.network_id = network_id
+            eth1_meta.updated_at = datetime.datetime.utcnow()
+        else:
+            self.logger.debug("New metadata: %s", remote)
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: Too noisy: `debug2`

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+                         network_id: int) -> None:
+        uri = remote.uri()
+        try:
+            node = self._get_remote(uri)
+        except NoResultFound:
+            self.logger.warning("Attempt to track metadata for untracked peer: %s", remote)
+            return
+
+        if self._meta_exists(uri):
+            self.logger.debug("Updated metadata: %s", remote)
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: Too noisy: `debug2`

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+
+    def track_peer_connection(self, remote: Node, is_outbound: bool) -> None:
+        uri = remote.uri()
+        if self._remote_exists(uri):
+            self.logger.debug("Updated peer: %s", remote)
+            node = self._get_remote(uri)
+            node.last_connected_at = datetime.datetime.utcnow()
+            return
+        else:
+            self.logger.debug("New peer: %s", remote)
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: too noisy: `debug2`

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+        else:
+            self.genesis_hash = genesis_hash
+        self.protocols = protocols
+        self.protocol_versions = protocol_versions
+        self.network_id = network_id
+
+    def track_peer_connection(self, remote: Node, is_outbound: bool) -> None:
+        uri = remote.uri()
+        if self._remote_exists(uri):
+            self.logger.debug("Updated peer: %s", remote)
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: too noisy: `debug2`

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+            *filters,
+        )
+        num_candidates = base_query.count()
+        self.logger.debug2(
+            "There are %d records in the peer tracking database",
+            num_candidates,
+        )
+        candidates = base_query.order_by(
+            Remote.last_connected_at.desc(),
+        ).all()[min(10, num_requested * 4 - 10):num_requested * 4]
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: TODO: remove this ugly hack.  The `get_peer_candidates` requests from the peer pool need to be updated to include the URIs for all of the peers the pool is currently connected to so that these backends can properly exclude them from their responses.

**Comment by pipermerriam** [comment]
Anyone who wants to re-connect to the network almost instantly can try this branch out.  It works well and my node comes online and starts syncing pretty much immediately

**Comment by pipermerriam** [comment]
I think I need help deciding the right way forward for this PR.  I *think* it's too much to do all in one go.  I'm thinking I'm going to drop the parts that source peer candidates from this API and have this PR only record the information, and do the sourcing of candidates in another PR.

Reasoning:

I'm concerned with not getting the sourcing right.  Here are some scenarios:

- Re-using the same ones too often
- Database full of mediocre peers 
- Database full of peers that were once available but no longer reachable
- Malicious/Incompetent peers in database that get blacklisted but are always eventually a valid connection candidate.

So probably inclined to release this behind two feature flags, one to turn it on at all, and another to enable sourcing of peer candidates from the recorded data.  Both of these would be off by default.

File: trinity/plugins/builtin/network_db/plugin.py [hunk 8]
```diff
@@ -151,3 +209,26 @@ def do_start(self) -> None:
...
+        else:
+            yield self._get_eth1_peer_server()
+
+    def do_start(self) -> None:
+        try:
+            tracker_services = self._get_services()
+        except BadDatabaseError as err:
+            self.logger.error("Shutting down trinity: %s", str(err))
+            # TODO: no easy way to make this work....  endpoints not yet connected.
+            # self.event_bus.request_shutdown(str(err))
```
Comment (cburgdorf) [inline trinity/plugins/builtin/network_db/plugin.py]: Just dropping this thought here: If you can catch this condition already in `def on_ready(self, manager_eventbus: TrinityEventBusEndpoint) -> None:` then you can use the `manager_eventbus` (that's the event bus from the `main` process` to propagate the shutdown.

**Comment by cburgdorf** [comment]
I just skimmed over this but responding to

>So probably inclined to release this behind two feature flags, one to turn it on at all, and another to enable sourcing of peer candidates from the recorded data. Both of these would be off by default.

I think that is the right way forward. In fact, I think we should leverage feature flags more in the future. Especially the more we stabilize, the more we probably need to get into the habit of letting some features  brew and fine tune behind feature flags for a while.

We could even try to come up with ways to promote such experimental features better, encouraging the adventurous to try them out (e.g. print out available feature flags for experimental features when Trinity boots).

File: p2p/peer.py [hunk 9]
```diff
@@ -274,9 +274,15 @@ def remove_subscriber(self, subscriber: 'PeerSubscriber') -> None:
             cmd, msg = await self.read_msg()
         if isinstance(cmd, Disconnect):
             msg = cast(Dict[str, Any], msg)
-            # Peers sometimes send a disconnect msg before they send the sub-proto handshake.
-            if msg['reason'] == DisconnectReason.too_many_peers.value:
-                raise TooManyPeersFailure(f'{self} disconnected from us before handshake')
+            try:
+                reason = DisconnectReason(msg['reason'])
```
Comment (pipermerriam) [inline p2p/peer.py]: TODO: extract these updates which better track/handle disconnect reasons.

File: p2p/peer_backend.py [hunk 10]
```diff
@@ -37,12 +38,16 @@ def __init__(self, event_bus: Endpoint) -> None:
 
     async def get_peer_candidates(self,
                                   num_requested: int,
-                                  num_connected_peers: int) -> Iterable[Node]:
+                                  connected_remotes: Set[Node]) -> Iterable[Node]:
```
Comment (pipermerriam) [inline p2p/peer_backend.py]: TODO: this API change can be extracted independently.

File: pytest.ini [hunk 11]
```diff
@@ -1,5 +1,5 @@
 [pytest]
-addopts= --showlocals --durations 50 --maxfail 10
+addopts= --showlocals --maxfail 10
```
Comment (pipermerriam) [inline pytest.ini]: revert

File: trinity/config.py [hunk 12]
```diff
@@ -246,7 +246,7 @@ def __init__(self,
                  port: int=30303,
                  use_discv5: bool = False,
                  preferred_nodes: Tuple[KademliaNode, ...]=None,
-                 bootstrap_nodes: Tuple[KademliaNode, ...]=None) -> None:
+                 bootstrap_nodes: Tuple[KademliaNode, ...]=None):
```
Comment (pipermerriam) [inline trinity/config.py]: TODO: revert

File: trinity/db/orm.py [hunk 13]
```diff
@@ -27,7 +27,7 @@
 Base = declarative_base()
 
 
-SCHEMA_VERSION = '2'
+SCHEMA_VERSION = '4'
```
Comment (pipermerriam) [inline trinity/db/orm.py]: TODO: revert to 3

File: trinity/plugins/builtin/network_db/connection/tracker.py [hunk 14]
```diff
@@ -57,10 +58,16 @@ def __init__(self, session: BaseSession):
     # Core API
     #
     async def record_blacklist(self, remote: Node, timeout_seconds: int, reason: str) -> None:
-        if self._record_exists(remote.uri()):
-            self._update_record(remote, timeout_seconds, reason)
+        try:
+            record = self._get_record(remote.uri())
+        except NoResultFound:
+            expires_at = datetime.datetime.utcnow() + datetime.timedelta(seconds=timeout_seconds)
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/connection/tracker.py]: TODO: this can be extracted to an independent PR

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 15]
```diff
@@ -0,0 +1,398 @@
...
+
+    eth1_meta = relationship("Eth1Meta", uselist=False, back_populates="remote")
+    blacklist = relationship(
+        "BlacklistRecord",
+        primaryjoin="Remote.uri==foreign(BlacklistRecord.uri)",
+        uselist=False,
+    )
+
+
+class Eth1Meta(Base):
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: I don't think I can justify these being two separate tables anymore.

TODO: collapse to a single table with combined fields.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 16]
```diff
@@ -0,0 +1,398 @@
...
+                         network_id: int) -> None:
+        pass
+
+    async def get_peer_candidates(self,
+                                  num_requested: int,
+                                  connected_remotes: Set[Node]) -> Iterable[Node]:
+        return ()
+
+
+MIN_QUALIFYING_UPTIME = 60
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: TODO: 2-5 minutes my be a better minimum.  1 minute is a very short connection.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 17]
```diff
@@ -0,0 +1,398 @@
...
+        uri = remote.uri()
+        now = datetime.datetime.utcnow()
+        if uptime < MIN_QUALIFYING_UPTIME:
+            # we only measure *meaningful* uptime.
+            uptime = 0
+
+        if self._remote_exists(uri):
+            self.logger.debug("Updated peer: %s", remote)
+            node = self._get_remote(uri)
+            node.last_connected_at = now if uptime >= MIN_QUALIFYING_UPTIME else None
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: TODO: this is wrong,  It shouldn't revert the value to `None` in this case, but rather preserve the previous value.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 18]
```diff
@@ -0,0 +1,398 @@
...
+            ~Remote.uri.in_(connected_uris),
+            *filters,
+        )
+        num_candidates = base_query.count()
+        self.logger.debug2(
+            "There are %d records in the peer tracking database",
+            num_candidates,
+        )
+        candidates = base_query.order_by(
+            Remote.last_connected_at.desc(),
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: TODO:  This function needs much bigger/fatter comment that explains the peer selection mechanism in more gory detail since it's going to effect node spin-up time in a major way and it's probably something we'll want to tweak.

File: p2p/peer_pool.py [hunk 19]
```diff
@@ -295,6 +295,8 @@ def _add_peer(self,
     async def stop_all_peers(self) -> None:
         self.logger.info("Stopping all peers ...")
         peers = self.connected_nodes.values()
+        for peer in peers:
+            peer.disconnect_reason = DisconnectReason.client_quitting
```
Comment (pipermerriam) [inline p2p/peer_pool.py]: I think this is wrong since it would be interpreted as the other client quitting.  Need to revert.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 20]
```diff
@@ -0,0 +1,400 @@
...
+)
+
+
+class Remote(Base):
+    __tablename__ = 'remotes'
+
+    id = Column(Integer, primary_key=True)
+
+    uri = Column(String, unique=True, nullable=False, index=True)
+    is_outbound = Column(Boolean, nullable=False, index=True)
```
Comment (pipermerriam) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: I tinkered with updating this to include extra concepts like `was_discovered` but it required contextual information that wasn't available and this PR is already large....  So I punted.

File: tests/core/p2p-proto/test_peer_metadata_tracker.py [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+    updated = tracker._get_remote(remote.uri())
+    assert updated.is_outbound is is_outbound
+
+
+def test_track_peer_connection_metadata(remote, caplog):
+    tracker = MemoryEth1PeerTracker()
+    tracker.track_peer_connection(remote, *TRACK_ARGS)
+
+    node = tracker._get_remote(remote.uri())
+    node = tracker._get_remote(remote.uri())
```
Comment (lithp) [inline tests/core/p2p-proto/test_peer_metadata_tracker.py]: this line is duplicated

File: tests/core/p2p-proto/test_peer_metadata_tracker.py [hunk 22]
```diff
@@ -0,0 +1,357 @@
...
+        tracker.track_peer_connection(remote, is_outbound, None, *meta_params)
+
+    for remote, is_outbound, meta_params in bad_remotes:
+        tracker.track_peer_connection(remote, is_outbound, None, *meta_params)
+
+    # use the same in-memory database
+    blacklist_tracker = SQLiteConnectionTracker(tracker.session)
+    for remote, delta_seconds in blacklist_records:
+        expires_at = datetime.datetime.utcnow() + datetime.timedelta(seconds=delta_seconds)
+        blacklist_tracker._create_record(remote, expires_at, 'test')
```
Comment (lithp) [inline tests/core/p2p-proto/test_peer_metadata_tracker.py]: for clarity it might be better to keep using `tracker` here rather than create `blacklist_tracker`.

File: tests/core/p2p-proto/test_peer_metadata_tracker.py [hunk 23]
```diff
@@ -0,0 +1,357 @@
...
+
+    candidates = tuple(sorted(
+        await tracker.get_peer_candidates(
+            num_requested=10,
+            connected_remotes=connected_remotes or set(),
+        ),
+        key=lambda r: r.uri(),
+    ))
+    just_good_remotes = tuple(r[0] for r in sorted(good_remotes, key=lambda r: r[0].uri()))
+    just_bad_remotes = tuple(r[0] for r in sorted(bad_remotes, key=lambda r: r[0].uri()))
```
Comment (lithp) [inline tests/core/p2p-proto/test_peer_metadata_tracker.py]: I don't think the rest of the test doesn't rely on these being sorted, if that's right I think removing that part would make the tests easier to read. Same with `candidates` above. Although another direction might be to lean into the sorting. I think the below 5 lines could be replaced with:

`assert candidates == just_good_results`

File: tests/core/p2p-proto/test_peer_metadata_tracker.py [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+            (NodeFactory(), True, SIMPLE_META),
+            (NodeFactory(), True, SIMPLE_META),
+        ),
+        (),
+    )
+
+
+@pytest.mark.asyncio
+async def test_getting_peer_candidates_excludes_non_outbound():
+    await do_tracker_peer_query_test(
```
Comment (lithp) [inline tests/core/p2p-proto/test_peer_metadata_tracker.py]: These calls are a bit hard to read. They might be easier to read if `tracker_params` was given a default of `{}`, since quite a few tests seem to do that. They also might be easier to read if the parameters were passed in by name:

```python
await do_tracker_peer_query_test(
    good_remotes=(
        (NodeFactory(), True, SIMPLE_META),
        (NodeFactory(), True, SIMPLE_META),
    ),
    bad_remotes=(
        (NodeFactory(), False, SIMPLE_META),
    ),
)
```

File: tests/core/p2p-proto/test_peer_metadata_tracker.py [hunk 25]
```diff
@@ -0,0 +1,357 @@
...
+            (NodeFactory(), True, (ZERO_HASH, 'les', 61, 1)),
+        ),
+        (
+            (NodeFactory(), True, (ZERO_HASH, 'bcc', 61, 1)),
+        ),
+    )
+
+
+@pytest.mark.asyncio
+async def test_getting_peer_candidates_excludes_protocol_version_mismatch():
```
Comment (lithp) [inline tests/core/p2p-proto/test_peer_metadata_tracker.py]: Another alternative for making these tests easier to parse:

```python
tracker = MemoryEth1PeerTracker(protocol_versions=(60, 61))

assert is_candidate(tracker, (NodeFactory(), True, (ZERO_HASH, 'eth', 60, 1)))
assert is_candidate(tracker, (NodeFactory(), True, (ZERO_HASH, 'eth', 61, 1)))
assert is_not_candidate(tracker, (NodeFactory(), True, (ZERO_HASH, 'eth', 62, 1)))
```

File: trinity/plugins/builtin/network_db/eth1_peer_db/server.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+    BaseEth1PeerTracker,
+)
+from .events import (
+    TrackPeerEvent,
+    GetPeerCandidatesRequest,
+    GetPeerCandidatesResponse,
+)
+
+
+class PeerDBServer(BaseService):
```
Comment (lithp) [inline trinity/plugins/builtin/network_db/eth1_peer_db/server.py]: Not a criticism of this PR but this is a 61 line file (and 48 lines of event definitions) which does nothing but expose two RPCs over the event bus. Hopefully a future version of lahja or maybe trinity will have better tooling which makes this easier!

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 27]
```diff
@@ -0,0 +1,400 @@
...
+    genesis_hash = Column(String, nullable=False, index=True)
+    protocol = Column(String, nullable=False, index=True)
+    protocol_version = Column(Integer, nullable=False, index=True)
+    network_id = Column(Integer, nullable=False, index=True)
+
+    blacklist = relationship(
+        "BlacklistRecord",
+        primaryjoin="Remote.uri==foreign(BlacklistRecord.uri)",
+        uselist=False,
+    )
```
Comment (lithp) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: The [SQLAlchemy docs for one-to-one relationships](https://docs.sqlalchemy.org/en/13/orm/basic_relationships.html#one-to-one) suggest using `back_populates` or `backref`. I'm not sure how that's different from the `primaryjoin` approach but wanted to point it out in case it was better in some way.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 28]
```diff
@@ -0,0 +1,400 @@
...
+        ExtendedDebugLogger,
+        logging.getLogger('trinity.protocol.common.connection.PeerTracker'),
+    )
+
+    msg_queue_maxsize = 100
+    subscription_msg_types: FrozenSet[Type[protocol.Command]] = frozenset()
+
+    def register_peer(self, peer: BasePeer) -> None:
+        # prevent circular import
+        from trinity.protocol.common.peer import BaseChainPeer
```
Comment (lithp) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: I never thought I would advocate for breaking things up into more files but this wouldn't be necessary if `BaseChainPeer` and `BaseChainPeerPool` were in different files.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 29]
```diff
@@ -0,0 +1,400 @@
...
+            last_connected_at=None,
+            genesis_hash=peer.genesis_hash,
+            protocol=peer.sub_proto.name,
+            protocol_version=peer.sub_proto.version,
+            network_id=peer.network_id,
+        )
+
+    def deregister_peer(self, peer: BasePeer) -> None:
+        """
+        This is likely the proper place to do record statistics about peer
```
Comment (lithp) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: `s/to do/to`

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py (hunk 16 above)
Comment (lithp) [reply trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: 1 minute seems reasonable for now. In the future we might care more about how useful the peer was than about how long it talked to us for, so leaving it this way sounds good for now!

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py [hunk 30]
```diff
@@ -0,0 +1,400 @@
...
+            # Either they have no blacklist record or the record is expired.
+            ((Remote.blacklist == None) | (BlacklistRecord.expires_at <= now)),  # noqa: E711
+            # We are not currently connected to them
+            ~Remote.uri.in_(connected_uris),  # type: ignore
+            # They match our filters for network metadata
+            *metadata_filters,
+        ).order_by(
+            # We want the ones that we have recently connected to succesfully to be first.
+            Remote.last_connected_at.desc(),  # type: ignore
+        ).all()
```
Comment (lithp) [inline trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: This probably won't matter for a while but once we know about a lot of nodes this might cause problems: `all()` apparently fetches everything and throws it into a list. [`yield_per`](https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.yield_per) looks like it might also not be what we want but I'm sure there's a way to read from a cursor and only have the database return the rows we want.

Comment (lithp) [reply trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: Alternatively, this method could use `num_requested`, which is currently unused.

File: trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py (hunk 27 above)
Comment (pipermerriam) [reply trinity/plugins/builtin/network_db/eth1_peer_db/tracker.py]: I don't remember all the things I tried but I believe that since the `BlacklistRecord.uri` is not a proper `ForeignKey` that the standard mechanisms ended up being in-adequate or not working in some cases.

File: tests/core/p2p-proto/test_peer_metadata_tracker.py (hunk 22 above)
Comment (pipermerriam) [reply tests/core/p2p-proto/test_peer_metadata_tracker.py]: These are two separate classes with different APIs.

File: trinity/plugins/builtin/network_db/eth1_peer_db/server.py (hunk 26 above)
Comment (pipermerriam) [reply trinity/plugins/builtin/network_db/eth1_peer_db/server.py]: I think a *lot* of this will get easier to slim down with the new lahja apis that are coming in since I think the new `EndpointAPI.subscribe` APIs are going to make it a lot easier for us *serve* endpoint responses without needing a `Service` to do it.

**Comment by veox** [comment]
But-but-but the CAP!..



Source: Yuri Smityuk / TASS via Getty Images via Vice

File: tests/core/p2p-proto/test_peer_metadata_tracker.py (hunk 22 above)
Comment (lithp) [reply tests/core/p2p-proto/test_peer_metadata_tracker.py]: oops, good point!
